### PR TITLE
Blog post draft on Z3 optimization checking

### PR DESCRIPTION
**Pull request by cfbolz**

### CODE REVIEW

File: posts/2022/12/pypy-alive.rst [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+.. tags:
+.. category:
+.. link:
+.. description:
+.. type: rest
+.. author: Carl Friedrich Bolz-Tereick
+
+In this blog post I want to describe a recent bug finding technique that I've
+added to the PyPy JIT testing infrastructure using the Z3 theorem prover. It's
+based on things I have learned from `John Regehr's`_ blog_, Twitter_, and on
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:13]: nit: it's a little confusing that "John Regehr's" and "blog" are separate links especially when by one of them you mean "blog post", I think ("blog" for me tends to mean the entire site)

File: posts/2022/12/pypy-alive.rst [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+.. _`John Regehr's`: https://www.cs.utah.edu/~regehr/
+.. _blog: https://blog.regehr.org/archives/1122
+.. _Twitter: https://twitter.com/johnregehr/
+.. __: https://www.cs.utah.edu/~regehr/alive2-pldi21.pdf
+
+Background: Bounds Analysis in the JIT
+======================================
+
+The optimizer of PyPy's JIT has an analysis based on `abstract interpretation`_
+that tries to find out, whether the integer values stored in a variable are
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:27]: nit: comma should not be there

File: posts/2022/12/pypy-alive.rst [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+actually not using the full 64 bit (or 32 bit) range, but instead fit into some
+smaller range. This means that for every integer variable ``x`` in a trace, the JIT
+compile tracks a range ``[a, b]`` with concrete integers ``a <= b`` in such a
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:30]: s/compile/compiler

File: posts/2022/12/pypy-alive.rst [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+way, that for every concrete runtime execution of the trace with a concrete
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:31]: I think you can rephrase this sentence and say something like "The JIT compiler tracks possible upper and lower bounds of the runtime value of a number: a range [a, b] such that for every concrete runtime value of some a number x, a <= x <= b"

File: posts/2022/12/pypy-alive.rst [hunk 5]
```diff
@@ -0,0 +1,727 @@
...
+value stored in ``x``, ``a <= x <= b`` must be true. ``a`` and ``b`` start out
+as the most general ``MININT`` and ``MAXINT``, but sometimes there is extra
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:33]: OT: how does it figure out if it can even bring the number into the machine word range anyway? does it just try auto unboxing everything?

File: posts/2022/12/pypy-alive.rst [hunk 6]
```diff
@@ -0,0 +1,727 @@
...
+information that makes it possible to improve these known bounds, and that is
+often useful to optimize the code.
+
+A typical example is that the JIT knows that the length of a string is
+non-negative, so for this kind of code: ``x = len(s)`` where ``s`` is a string,
+``x`` gets a range ``[0, MAXINT]`` assigned. With this information we could for
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:39]: Do you allow strings that are MAXINT long?

File: posts/2022/12/pypy-alive.rst [hunk 7]
```diff
@@ -0,0 +1,727 @@
...
+example remove a check ``x + 10 < 0`` completely, because it can never be True.
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:40]: backticks around `True`

File: posts/2022/12/pypy-alive.rst [hunk 8]
```diff
@@ -0,0 +1,727 @@
...
+
+The bounds information is very useful for optimization, but the analysis of the
+bounds is also a source of bugs in the JIT, because the reasoning is often
+subtle and easy to get wrong in corner cases. We already use a number of testing
+techniques to try to make sure that it is correct. A simple one is
+`property-based testing`_ using Hypothesis_ on the operations on bounds.
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:46]: needs a transition sentence like "but that's not always enough" (ooh spooky)

File: posts/2022/12/pypy-alive.rst [hunk 9]
```diff
@@ -0,0 +1,727 @@
...
+
+I am currently supervising a Bachelor thesis by Nico Rittinghaus, who is
+extending the integer analysis in the JIT. He'll probably write a separate blog
+post about it soon. In the process of doing that, the current bounds analysis
+code got a lot of scrutiny, and we found out that one of the unit tests of the
+bounds analysis was actually incorrect, and the example code in that unit test
+was optimized incorrectly. This case of incorrect optimization is not a big deal
+for regular Python code, because it involved a "wrapping integer addition
+operation", i.e. one where overflowing results just wrap around to negative
+values. In Python, all additions actually have overflow checks and convert to
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:63]: I think this sentence is a little funky because Python says no such things about overflow and machine words. That's an implementation detail. IMO rewrite to suggest that Python integers are theoretically of unbounded size, etc

File: posts/2022/12/pypy-alive.rst [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+a big integer representation if the results don't fit into a machine word.
+However, it's still possible to trigger the problem with the special
+``__pypy__.intop.int_add`` API which is a function that exposes wraparound
+arithmetic on ints.
+
+Here's the miscompilation. The JIT optimizes the following function:
+
+https://foss.heptapod.net/pypy/pypy/-/issues/3832
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:71]: This link feels a little out of place without a label or something

File: posts/2022/12/pypy-alive.rst [hunk 11]
```diff
@@ -0,0 +1,727 @@
...
+
+    def wrong(x):
+        a = __pypy__.intop.int_add(x, 10)
+        if a < 15:
+            return 0
+        return 2
+
+Basically the faulty reasoning of the JIT looks like this: if ``int_add(x, 10) < 15``
+then it must follow that ``x < 5``, which is stronger than ``x < 6``, so the
+second ``if`` is always True. This sounds good, but is of course actually wrong
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:100]: nit: remove "of course" (and "obviously" from other places in the article)

File: posts/2022/12/pypy-alive.rst [hunk 12]
```diff
@@ -0,0 +1,727 @@
...
+if the addition ``+ 10`` wrapped around. So if ``x == MAXINT``, then
+``int_add(x, 10) == MININT + 9 < 15``. But ``MAXINT < 5`` is of course not
+correct.
+
+Note how the same reasoning with overflow-checking addition is correct! If ``x +
+10 < 15`` and the ``+`` didn't overflow, then indeed ``x < 6``. And if your
+brain bends starting to think about all this, you understand some of the
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:107]: I think mind-bending is the idiom

File: posts/2022/12/pypy-alive.rst [hunk 13]
```diff
@@ -0,0 +1,727 @@
...
+difficulty of getting the JIT correct in this area.
+
+How could we have avoided this bug?
+=====================================
+
+One `exercise I try to do after finding bugs`_ is to reflect on ways that the
+bug could have been avoided. I think this is particularly important in the JIT,
+where bugs are potentially really annoying to find.
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:115]: I think there's also an unstated "...and cause extremely bizarre second- and third-order effects"

File: posts/2022/12/pypy-alive.rst [hunk 14]
```diff
@@ -0,0 +1,727 @@
...
+to get wrong, so probably they should never have been introduced in the first
+place (and my eventual bugfix will indeed remove them).
+
+But in the rest of this blog post I want to talk about a very systematic
+approach that can be applied to the problem of mis-optimizations of integer
+operations, and that is done by applying an SMT solver to the problem.
+
+An SMT solver (`Satisfyability Modulo Theories`_) is a tool that can be used to
+find out whether a certain mathematical formula is "satisfiable", i.e. whether
+some chosen set of input will make the formula evaluate to True. SMT solvers are
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:134]: backticks on True?

File: posts/2022/12/pypy-alive.rst [hunk 15]
```diff
@@ -0,0 +1,727 @@
...
+described on his blog.
+
+.. __: https://github.com/AliveToolkit/alive2/
+.. __: https://kristerw.github.io/2022/09/13/translation-validation/
+
+
+Z3 Proof of Concept
+=======================
+
+The first thing I did was to try to get Z3 find the above bug, by encoding the
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:160]: OT: I think lifting other tools into languages & compilers is very cool and very underrated

File: posts/2022/12/pypy-alive.rst [hunk 16]
```diff
@@ -0,0 +1,727 @@
...
+    cond1 = a < 15
+    cond2 = x < 6
+    prove(Implies(cond1, cond2))
+
+
+Here, ``x`` is defined to be a bit vector variable of width 64, which is a
+datatype that can be used to represent a bounded machine integers. Addition on
+bit vectors performs wraparound arithmetic, like the ``__pypy__.intop.int_add``
+call in the original code. The JIT optimized the second condition away, so
+essentially it was convinced that the first condition implies the second one. So
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:178]:
```suggestion
essentially it was convinced that the first condition implies the second one. The
```

File: posts/2022/12/pypy-alive.rst [hunk 17]
```diff
@@ -0,0 +1,727 @@
...
+
+When run, the above program prints::
+
+    counterexample
+    [x = 9223372036854775803]
+
+Which shows the bug. As a small side-note, I thought it was cool that the
+process of "proving" something in Z3 basically means trying to find an example
+for the negation of the formula. If no counterexample can be found for the
+negation, the original formula is True. If the original formula turns out to be
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:189]: Is that guaranteed? Or is there some other, third thing where it goes on computing forever?

File: posts/2022/12/pypy-alive.rst [hunk 18]
```diff
@@ -0,0 +1,727 @@
...
+
+- After a guard, we tell Z3 that from now on it can assume that the guard
+  condition is True.
+
+- We repeat this, guard for guard, until we reach the end of the trace. There,
+  we ask Z3 to prove that the output variables in the unoptimized trace and the
+  optimized trace are identical (every trace can return one or many values).
+
+I implemented this, it's `not a lot of code`__, basically a couple of hundred lines
+of (not particularly great) Python code. So far I only support integer
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:237]: no need to talk yourself down here. remove the parenthetical

File: posts/2022/12/pypy-alive.rst [hunk 19]
```diff
@@ -0,0 +1,727 @@
...
+           ...
+            # convert arguments
+            if op.numargs() == 1:
+                arg0 = self.convertarg(op, 0)
+            elif op.numargs() == 2:
+                arg0 = self.convertarg(op, 0)
+                arg1 = self.convertarg(op, 1)
+
+            # compute results
+            if opname == "int_add":
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:262]: do you have an `operator` module?

File: posts/2022/12/pypy-alive.rst [hunk 20]
```diff
@@ -0,0 +1,727 @@
...
+
+Some things are a little bit trickier. An important example in the context of
+this blog post are integer operations that check for overflow. The overflow
+operations return a result, but also a boolean whether the operation overflowed
+or not.
+
+.. code:: python
+
+    def add_to_solver(self, ops, state):
+            ...
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:383]: imo add `#` before `...`

File: posts/2022/12/pypy-alive.rst [hunk 21]
```diff
@@ -0,0 +1,727 @@
...
+
+            # more cases
+            elif opname == "int_add_ovf":
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:386]: it might be worth noting that this kind of technique is also used by some of the superopt work to model x86, etc

File: posts/2022/12/pypy-alive.rst [hunk 22]
```diff
@@ -0,0 +1,727 @@
...
+addition case step by step, the other cases work analogously.
+
+The addition of the ``elif`` that computes ``expr`` is an addition on bit
+vectors, therefore it is performing wraparound arithmetic.
+``z3.SignExt(LONG_BIT, arg0)`` sign-extends ``arg0`` from a bit vector of
+``LONG_BIT`` bits to an abstract, arbitrary precision integer. The addition in
+the second line is therefore an addition between abstract integers, so it will
+never overflow and just compute the correct result as an integer.
+
+The condition to check for overflow is now: Overflow did
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:411]: put the "if" first since you are saying "the condition is"

File: posts/2022/12/pypy-alive.rst [hunk 23]
```diff
@@ -0,0 +1,727 @@
...
+not occur if the results of the two different ways to do the addition are the
+same. So in the third line the code converts the result of the bit vector
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:413]: "the third line" is a little ambiguous and could use re-printing or highlight the line above or something

File: posts/2022/12/pypy-alive.rst [hunk 24]
```diff
@@ -0,0 +1,727 @@
...
+
+        elif opname == "guard_no_overflow":
+            assert state.no_ovf is not None
+            return state.no_ovf
+        elif opname == "guard_overflow":
+            assert state.no_ovf is not None
+            return z3.Not(state.no_ovf)
+        ...
+
+Finding the Bug, Again
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:433]: I think this section is a little hard to follow since there is a bunch of back and forth between different traces

File: posts/2022/12/pypy-alive.rst [hunk 25]
```diff
@@ -0,0 +1,727 @@
...
+    i15 = int_lt(i8, i4) # 1
+    i16 = int_and(i6, i0) # 8
+    i17 = uint_ge(i6, -6/ci) # 0
+    finish()
+
+Note how every guard generated is true for the example values.
+
+I have been running this combination of random trace generation and Z3 checking
+for many nights and it has found some bugs! It should probably could still be
+run for a lot longer, but still a very useful exercise already.
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:624]: Maybe you can run it as part of CI for N hours per commit and have it report its findings somewhere

Comment (tekknolagi) [reply posts/2022/12/pypy-alive.rst:624]: A pseudo cron job

File: posts/2022/12/pypy-alive.rst [hunk 26]
```diff
@@ -0,0 +1,727 @@
...
+
+Bugs Found
+============
+
+In addition to the two bugs I've already described, here are the further bugs
+that were found, all not really by Z3, but because asserts were triggered:
+
+- The JIT concluded after ``12 == int_mul(x, 12)`` that ``x == 1``, which is
+  incorrect if overflow occurred.
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:633]: maybe provide counterexamples since that's what z3 does?

File: posts/2022/12/pypy-alive.rst [hunk 27]
```diff
@@ -0,0 +1,727 @@
...
+    finish(r)
+
+This was optimized to:
+
+.. code:: python
+
+    [a, b]
+    finish(a)
+
+Which is incorrect, because the guard can obviously fail given the right inputs.
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:669]: remove "obviously"

File: posts/2022/12/pypy-alive.rst [hunk 28]
```diff
@@ -0,0 +1,727 @@
...
+doing, for example). As of now, the code is on a not-yet-merged branch__, I plan
+to integrate it into our testing infrastructure after the next PyPy release is
+out.
+
+.. __: https://foss.heptapod.net/pypy/pypy/-/tree/branch/fix-intutils-ovf-bug
+
+Acknowledgements
+=================
+
+Thanks to Saam Barati and Joshua Schmidt for great feedback on drafts of this post!
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:727]: Do they have websites to link?

**Review by tekknolagi — COMMENTED** [review]
Looking good! Some comments

File: posts/2022/12/pypy-alive.rst (hunk 1 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:13]: done, now I link to both

File: posts/2022/12/pypy-alive.rst (hunk 4 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:31]: went with a slight variant

File: posts/2022/12/pypy-alive.rst (hunk 5 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:33]: that's kind of beyond the scope of the post and orthogonal to the issues the post is talking about. But I'll add a paragraph saying a little bit about it, because other people probably have a similar question.

the answer is: it doesn't unbox all integer variables. it only unboxes those where the concrete runtime value seen during tracing does fit into a machine word. this is also how pypy3 works internally, there are two different representations for `int`, one using machine ints and one using the equivalent of `longobject.c`. Basically it's like Python2 used to work, except only on the implementation level.

File: posts/2022/12/pypy-alive.rst (hunk 6 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:39]: if you have enough ram, sure! you might, on 32bit machines for example.

File: posts/2022/12/pypy-alive.rst (hunk 15 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:160]: agreed!

File: posts/2022/12/pypy-alive.rst (hunk 17 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:189]: it will always terminate, but yes, you might not be willing to wait long enough. indeed, I am running Z3 with a time limit, will write a little bit about it somewhere.

File: posts/2022/12/pypy-alive.rst (hunk 18 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:237]: I'll write "somewhat hacky" instead, because it definitely is that!

File: posts/2022/12/pypy-alive.rst (hunk 21 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:386]: what do you mean by "This kind of technique"? Z3 in general?

File: posts/2022/12/pypy-alive.rst (hunk 19 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:262]: you mean to be able to write some of the cases this way?:

```python
            if opname.startswith("int_") and hasattr(operator, opname[4:]):
                expr = getattr(operator, opname[4:])(arg0, arg1)
```

or do you mean "do we have an equivalent module to operator that lists all IR operations"?

File: posts/2022/12/pypy-alive.rst (hunk 9 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:101]: I'm changing it to "All the additions that the Python frontend generates actually have overflow checks"

File: posts/2022/12/pypy-alive.rst (hunk 24 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:484]: yeah, I radically simplified the trace snippets in this section (only one figure left for the traces)

File: posts/2022/12/pypy-alive.rst (hunk 9 above)
Comment (tekknolagi) [reply posts/2022/12/pypy-alive.rst]: Python is abstract

File: posts/2022/12/pypy-alive.rst [hunk 29]
```diff
@@ -0,0 +1,769 @@
...
+To given some wider context I'll explain a bit how Python ints in the user code
+relate to the types that are used in traces.
+When turning a regular Python 3 function into a trace, there is a lot of work
+happening in the JIT frontend to try to observe and infer the types that the
+Python function concretely uses at runtime. The traces are generated under these
+typing assumptions. Therefore, code that uses ``ints`` in the Python code can
+typically be translated into traces that operate on machine integers. In order
+to make sure that the Python integer semantics are upheld, many of the
+operations in the traces need to check that the integer results of some
+operations still fit into a machine integer. If that is not the case (a rare
```
Comment (tekknolagi) [inline posts/2022/12/pypy-alive.rst:44]: new section looks good to me

**Review by tekknolagi — APPROVED** [review]
looking good

File: posts/2022/12/pypy-alive.rst (hunk 19 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:304]: you can but it only works for a subset of the operations and it would be full of special cases (oh, it's `int_and`! nope, `operator.and` is not a thing of course, so you need `operator.and_`, same for or and not). I think just listing the handful of actual cases is fine.

File: posts/2022/12/pypy-alive.rst (hunk 21 above)
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:431]: I don't quite feel qualified to point out what other cool things are done with Z3 in the compiler space, unfortunately :-(

File: posts/2022/12/pypy-alive.rst [hunk 30]
```diff
@@ -0,0 +1,769 @@
...
+instructions with guards. The instructions in the traces operate on different
+data types, machine integers, doubles, pointers, bools, etc. In this post we'll
+be mostly concerned with machine integers.
+
+To given some wider context I'll explain a bit how Python ints in the user code
+relate to the types that are used in traces.
+When turning a regular Python 3 function into a trace, there is a lot of work
+happening in the JIT frontend to try to observe and infer the types that the
+Python function concretely uses at runtime. The traces are generated under these
+typing assumptions. Therefore, code that uses ``ints`` in the Python code can
```
[the remark this answers is not in the document]
Comment (cfbolz) [reply posts/2022/12/pypy-alive.rst:41]: I don't quite like it either but I'll leave it, restructuredtext doesn't actually support doing this properly, I think, and I can't be bothered to hack around it.

File: posts/2022/12/pypy-alive.rst (hunk 19 above)
Comment (tekknolagi) [reply posts/2022/12/pypy-alive.rst:306]: you need `and_` because `and` is a kw

File: posts/2022/12/pypy-alive.rst (hunk 21 above)
Comment (tekknolagi) [reply posts/2022/12/pypy-alive.rst:433]: ¯\_(ツ)_/¯

someone else will inevitably point it out in the comments
